### ISSUE TITLE
Stutter in video playback when changing volume or bringing up OSD

### ISSUE THREAD

**Issue by theultimateinferno** [jurialmunkey/skin.arctic.horizon.2#650]
### Skin section

Fullscreen Video OSD

### Current Behavior

Upon changing volume or opening up OSD slight stutter in video playback. Did not occur on previous versions of skin or in Estuary. Also happens when streaming trailers from YouTube. 

### Expected Behavior

No stutter. 

### Steps To Reproduce

Video playback, trailer playback. Happens on USB storage, nas and streaming. 

### Checklist

- [X] I have made a separate issue for each bug that I am reporting.
- [X] I have searched the issues and this bug has not been reported.
- [X] I am using the most recent skin version.
- [X] I have confirmed that this bug does not occur in the default skin.

**Comment by jurialmunkey** [comment]
I don't experience this issue and it could be caused by any number of things, so it is difficult for me to know what.

If you can pin-point exactly which version it started with then I can look at what was added in that version and go from there. Without more information I'm completely in the dark.

**Comment by Ozymandyaz** [comment]
I have this issue also and as far as I can tell its been that way for a while (a few months?) That being said, I think its a CPU constraint and not a bug per se... I have a Raspberry Pi 400 and when I increase the clock speed I see less of a stutter but I am not able to get it to go away completely. My suspicion is that the complexity of the OSD causes a CPU spike when the OSD is called, so low power devices may show the issue where faster players may not. 

I can also see that the stutter temporarily goes away after the first time the OSD is called. That is to say, the first time the OSD is called, there is a stutter. A few seconds later, the OSD can be called with no stutter. But if I wait a few minutes the stutter will come back. Seems like Kodi caches some of the skin elements, so the second time they are called it goes faster.

**Comment by theultimateinferno** [comment]
> I have this issue also and as far as I can tell its been that way for a while (a few months?) That being said, I think its a CPU constraint and not a bug per se... I have a Raspberry Pi 400 and when I increase the clock speed I see less of a stutter but I am not able to get it to go away completely. My suspicion is that the complexity of the OSD causes a CPU spike when the OSD is called, so low power devices may show the issue where faster players may not.
> 
> I can also see that the stutter temporarily goes away after the first time the OSD is called. That is to say, the first time the OSD is called, there is a stutter. A few seconds later, the OSD can be called with no stutter. But if I wait a few minutes the stutter will come back. Seems like Kodi caches some of the skin elements, so the second time they are called it goes faster.

Thanks, this makes sense to me and sounds the same as my issue. I've been back and tried the last 10 releases but can't isolate when it started occurring. If it's a hardware issue then perhaps nothing can be done. I can live with it, not a major deal!

**Comment by jurialmunkey** [comment]
Yeah on a very low powered device like an rpi I wouldn't be surprised about the osd. It almost certainly would be the playlist/cast (ondown) and audio/subs lists loading. They only need to load once, which is why you wouldn't see it again reopen later.

The volume control causing a stutter is weird though as it isn't that complex.

**Comment by jurialmunkey** [comment]
Yeah on a very low powered device like an rpi I wouldn't be surprised about the osd. It almost certainly would be the playlist/cast (ondown) and audio/subs lists loading. They only need to load once, which is why you wouldn't see it again on a  reopen later.

The volume control causing a stutter is weird though as it isn't that complex.

**Comment by theultimateinferno** [comment]
> Yeah on a very low powered device like an rpi I wouldn't be surprised about the osd. It almost certainly would be the playlist/cast (ondown) and audio/subs lists loading. They only need to load once, which is why you wouldn't see it again on a reopen later.
> 
> The volume control causing a stutter is weird though as it isn't that complex.

I'm running AH2 on a Sony OLED A8 which is quite powerful so I don't quite understand why its happening. The volume change issue is perplexing to me!

**Comment by jbelew** [comment]
Chiming in here, I'm also starting to see just the slightest bit of video frame drops when adjusting the audio offset on my oDroid N2+. Seems to be a regression starting around 0.7,6. As for the RP4, it's not a bad little box, especially if you put a high-quality SD card in it. It's a pure Linux install so there's none of the unpredictability of Android that's causing all the noise on the forums. If you ever get to the point of going back through the skin for an optimization pass, it would be a good platform to target/test on.

**Comment by theultimateinferno** [comment]
> Chiming in here, I'm also starting to see just the slightest bit of video frame drops when adjusting the audio offset on my oDroid N2+. Seems to be a regression starting around 0.7,6. As for the RP4, it's not a bad little box, especially if you put a high-quality SD card in it. With LibreElec, it'll do 12-bit 4k / HDR. It's also a pure Linux install so there's none of the unpredictability of Android that's causing all the noise on the forums. If you ever get to the point of going back through the skin for an optimization pass, it would be a good platform to target/test on.

Thanks. Out of interest, do you get this occurring on other skins?

**Comment by jurialmunkey** [comment]
Can you guys try v0.7.8 and see if that helps with the OSD stutters?

I've moved the OSD lists out into dedicated dialogs, so that should cut down on the amount of controls that load when the videoosd is first opened.

**Comment by Ozymandyaz** [comment]
> 

Just tried on my Pi400 - it’s a bit better for sure, but there is still a slight stutter when the OSD loads.

**Comment by jurialmunkey** [comment]
> Just tried on my Pi400 - it’s a bit better for sure, but there is still a slight stutter when the OSD loads.

Is this only videoosd controls or also just the seekbar by itself? Do you have a remote with an info button to just pull up the seekbar without the controls? 

Possibly could be a texture loading issue.

**Comment by theultimateinferno** [comment]
> > Just tried on my Pi400 - it’s a bit better for sure, but there is still a slight stutter when the OSD loads.
> 
> Is this only videoosd controls or also just the seekbar by itself? Do you have a remote with an info button to just pull up the seekbar without the controls?
> 
> Possibly could be related to texture loading for the background fade as it is quite large.

Thanks for looking into this, I will test my set up tonight. Not sure I have that button on remote but can plug in a keyboard, what would the shortcut be?

**Comment by jurialmunkey** [comment]
Usually is mapped to `i` 

It's just a way to bring up the seekbar without pausing or skipping and without the videoosd controls. That we can determine if the stutter is caused by something specifically in the videoosd, or if it is something in the seekbar (or hopefully it is gone altogether now).

**Comment by theultimateinferno** [comment]
0.7.8 is significantly better. There is a very small stutter on opening OSD or pressing 'i' however hardly noticeable now unless you are looking for it. Stutter on volume change appears to have gone away. Thank you so much.

**Comment by jweinr** [comment]
For me it's been like this since AH1 on my shield tv 2017 running a 4K interface.
Usually on the first bringing up of the osd, then it's fine for a while. Then it will stutter again.
There's actually frame skipping logged in the debug when this happens.
I always figured it was because of AH being heavier and the CPU being too weak and never gave it much thought.
Still the same on the latest version.

**Comment by illuvattarr** [comment]
I'm alsp experiencing this exact issue on vero 4k+ on version 0.7.14. 

The stutter happens both when seeking and when brining up the OSD options. Then when bringing it up again right away there is no stutter. It only happens again after some time passed.

**Comment by jurialmunkey** [comment]
It still sounds like it could possibly be Texture loading issues. Let's try packing the textures into a Textures.xbt for a bit better performance on that side of things and see if it helps

Download the Textures.xbt file from here:
https://github.com/jurialmunkey/repository.jurialmunkey/raw/master/resources/skin.arctic.horizon.2/Textures.xbt

Add the file to `addons/skin.arctic.horizon.2/media` and restart Kodi.

NOTE: The file will get overridden everytime you update the skin. If it solves the issue I will look at a more permenant solution to add the Textures.xbt with each update but for now I want to make sure it actually does resolve the issue before I go to the trouble.

**Comment by jweinr** [comment]
I think you're on the right track.
On my STV2017:
Just tested the packaged Textures.xbt, 
There's still a visible stutter, but much shorter now.
No frames are skipped now also, while there used to skipped 1 or 2 frames before.

On an unrelated topic, are you aware of this artifact when navigating the osd? Been like this since the first version of AH2.
https://photos.app.goo.gl/nJHq66SU4ze6XoReA

Just wanna know if I should submit a report or not.

Thank you

**Comment by theultimateinferno** [comment]
> 

I have tried this on Sony A8. This eliminates the stutter issue for me!

**Comment by jurialmunkey** [comment]
Thanks.

@jweinr that flicker is a kodi rendering issue not the skin. There's not a lot I can do if it is struggling to render the texture change from unfocused to focus state fast enough that it's causing the controls to jump.

**Comment by jweinr** [comment]
Got it, thanks.

Jan Weinreich Schjolberg
Gerente Comercial
Grupo Delsantek
+56 9 7391 4444

**Comment by illuvattarr** [comment]
Just tried it as well on my Vero 4K+ and it's definitely improved. Not 100% sure it's totally gone, but it's for sure less noticeable. I will try some more testing later today.

**Comment by gonzolak** [comment]
This file works like a charm for OSD on my setup. (libreelec, intel i5 , integrated gfx). What I noticed also improved animation stutter!!!! It was bugging my for so long. When I play video or live tv and then go back to main screen animation are 100% fluid! So it's like you said it's redering issue. When I don't play video or live tv it still stutter thought :( Is this xbt file workaround only work when video is playing??? Can you make it work when no video is playing on main screen?? That would be great.

**Comment by jurialmunkey** [comment]
@gonzolak - the textures file contains all the textures from the media folder. It will affect the entire skin not just video.

However there are some textures that must stay outside the file because they either need to be selectable (eg icons for the menu) or are generated dynamically (eg the gradient highlight)

**Comment by Ozymandyaz** [comment]
The file also improved my setup. For my Pi 400 I can overclock to 2200mhz and use force_turbo=1 to keep the clock running at max all the time. With this Pi OC config and the XBT file the stutter is almost gone, its like a couple dropped frames that are barely noticeable.

**Comment by jweinr** [comment]
Still relevant and the textures file useful on Nexus.

**Comment by jurialmunkey** [comment]
@theultimateinferno (and everyone else with this issue) 

Please try v0.8.6 and *disable* this setting:
`Skin Settings > Behaviour > Auto pop-up OSD dialogs`

Based upon a comment from a forum user I think possibly some of these stutters are from the skinvariables checking the audio/subtitle streams for the pop-up dialogs on the right side buttons.

Disabling this setting will move the audio/subs streams into a separate dialog outside the videoosd -- which means it will delay loading until it is called. It will be slightly inconvenient in that you will need to press up to access the audio/subs selector rather than it popping up automatically but I think it could help reduce the initial stutter in the osd.

**Comment by theultimateinferno** [comment]
> @theultimateinferno (and everyone else with this issue)
> 
> Please try v0.8.6 and _disable_ this setting: `Skin Settings > Behaviour > Auto pop-up OSD dialogs`
> 
> Based upon a comment from a forum user I think possibly some of these stutters are from the skinvariables checking the audio/subtitle streams for the pop-up dialogs on the right side buttons.
> 
> Disabling this setting will move the audio/subs streams into a separate dialog outside the videoosd -- which means it will delay loading until it is called. It will be slightly inconvenient in that you will need to press up to access the audio/subs selector rather than it popping up automatically but I think it could help reduce the initial stutter in the osd.

Thanks. Tried this version and turning off that setting but there is still a very small stutter for me It's only on the first loading of the OSD. I'm used to it and it's not a huge issue

**Comment by illuvattarr** [comment]
There's still a small stutter after disabling the setting. The only thing that fixes it for me is adding the Textures file, which is fine by me. Any way to retain that file through auto updates perhaps?

**Comment by jweinr** [comment]
Disabled "Auto pop-up OSD dialogs" and still get a stutter on latest version. The textures files still improves things.

I'm up for testing more extreme solutions / builds of AH2.

**Comment by jurialmunkey** [comment]
> Disabled "Auto pop-up OSD dialogs" and still get a stutter on latest version. The textures files still improves things.

I'm fairly sure it is primarily a texture loading issue. The auto-pop was a long shot. 

Reasoning:

Kodi has two methods for an image control to load textures: sync and background. 

**Sync Loading**

- Texture loaded in the main GUI thread. 
- If Kodi cannot load the texture fast enough, it will hold up the GUI thread and cause jankiness due to the frame drops
- Textures.xbt helps loading speed because it preconverts PNGs into the same format as the renderer

**Background Loading**
- Texture loaded in a separate background thread to avoid jankiness
- If Kodi cannot load the texture fast enough, the texture will be blank and then pop-in once it has loaded
- Suitable for artwork since it isn't critical to display immediately. Less suitable for interface elements like OSD buttons.

**AH2 Problem**
- AH2 has interface textures which cannot be preconverted into an XBT
  - Icons because you need to be able to browse them in skinshortcuts to select them
  - Gradients because they are dynamically generated via a script based on user's colours

> I'm up for testing more extreme solutions / builds of AH2.

@jweinr - If you're looking for a more extreme solution, you could try compiling the icons and gradients into the Textures.xbt

You can find a GUI tool for windows to convert textures here:
https://kodi.wiki/view/TextureTool

The Icons and Gradients can be found in these folders
`kodi/addons/skin.arctic.horizon.2/extras/icons`
`kodi/userdata/addon_data/script.texturemaker/ArcticHorizon`

You should move the PNGs from them into the skin's media folder as separate folders e.g.
`kodi/addons/skin.arctic.horizon.2/media/test/icons`
`kodi/addons/skin.arctic.horizon.2/media/test/gradients`

After moving the textures, use the TextureTool to compress the media folder into a Textures.xbt
Then use a text editor to find and replace all instances of those paths in the xml files in 1080i

e.g. 
Find:
`special://profile/addon_data/script.texturemaker/ArcticFuse/`
Replace With:
`test/gradients/`

Find:
`special://skin/extras/icons/`
Replace with:
`test/icons/`

Then restart Kodi.

You're going to have to do this everytime you update the skin or you want to change colours, so it is a very extreme method. However, if it does turn out to make a big difference then I consider some changes to the skin to make it a bit easier to apply as a patch -- I just don't want to uproot the whole skin if it's not actually going to significantly change things.

**Comment by Bindou01** [comment]
Are you sure this is not related to the display frequency of the interface (60hz) VS the native frequency of the video you are playing (24hz) ? When the OSD appears, the output recalibrates to the frequency of the interface rather than the film.

**Comment by jurialmunkey** [comment]
> Are you sure this is not related to the display frequency of the interface (60hz) VS the native frequency of the video you are playing (24hz) ? When the OSD appears, the output recalibrates to the frequency of the interface rather than the film.

This shouldn't happen. Kodi will only change the display frequency when you leave fullscreen video (e.g. go back to home). It won't change when opening video OSD controls.

**Comment by basicals** [comment]
> > Disabled "Auto pop-up OSD dialogs" and still get a stutter on latest version. The textures files still improves things.
> 
> I'm fairly sure it is primarily a texture loading issue. The auto-pop was a long shot.
> 
> Reasoning:
> 
> Kodi has two methods for an image control to load textures: sync and background.
> 
> **Sync Loading**
> 
> * Texture loaded in the main GUI thread.
> * If Kodi cannot load the texture fast enough, it will hold up the GUI thread and cause jankiness due to the frame drops
> * Textures.xbt helps loading speed because it preconverts PNGs into the same format as the renderer
> 
> **Background Loading**
> 
> * Texture loaded in a separate background thread to avoid jankiness
> * If Kodi cannot load the texture fast enough, the texture will be blank and then pop-in once it has loaded
> * Suitable for artwork since it isn't critical to display immediately. Less suitable for interface elements like OSD buttons.
> 
> **AH2 Problem**
> 
> * AH2 has interface textures which need to be sync loaded but cannot be preconverted into an XBT
>   
>   * Icons because you need to be able to browse them in skinshortcuts to select them
>   * Gradients because they are dynamically generated via a script based on user's colours
> 
> > I'm up for testing more extreme solutions / builds of AH2.
> 
> @jweinr - If you're looking for a more extreme solution, you could try compiling the icons and gradients into the Textures.xbt
> 
> You can find a GUI tool for windows to convert textures here: https://kodi.wiki/view/TextureTool
> 
> The Icons and Gradients can be found in these folders `kodi/addons/skin.arctic.horizon.2/extras/icons` `kodi/userdata/addon_data/script.texturemaker/ArcticHorizon`
> 
> You should move the PNGs from them into the skin's media folder as separate folders e.g. `kodi/addons/skin.arctic.horizon.2/media/test/icons` `kodi/addons/skin.arctic.horizon.2/media/test/gradients`
> 
> After moving the textures, use the TextureTool to compress the media folder into a Textures.xbt Then use a text editor to find and replace all instances of those paths in the xml files in 1080i
> 
> e.g. Find: `special://profile/addon_data/script.texturemaker/ArcticHorizon/` Replace With: `test/gradients/`
> 
> Find: `special://skin/extras/icons/` Replace with: `test/icons/`
> 
> Then restart Kodi.
> 
> You're going to have to do this everytime you update the skin or you want to change colours, so it is a very extreme method. However, if it does turn out to make a big difference then I can consider some changes to the skin to make it a bit easier to apply as a patch -- I just don't want to uproot the whole skin if it's not actually going to significantly change things.

Allright followed your instructions to the letter.
i'm getting what seems like corrupted or missing textures.

![image](https://github.com/jurialmunkey/skin.arctic.horizon.2/assets/6373803/38253af5-84bb-464c-b3eb-66aa2a3012d1)


And the frame skipping is still there, as subtle as when using the xbt file you shared previously.
